packages/aila/src/lib/agents/promptAgentHandler.ts

```diff
@@ -19,54 +25,150 @@ export async function promptAgentHandler<Schema extends SchemaWithValue>({
+  agent: PromptAgentDefinition<Schema, SchemaStrict>;
   document: LooseLessonPlan;
   targetKey: LessonPlanKey;
   additionalInstructions: string;
   chatId: string;
   userId: string;
   ragData: CompletedLessonPlan[];
-}): Promise<{ content: z.infer<Schema>["value"] }> {
+}): Promise<
+  | { success: true; content: z.infer<SchemaStrict> }
+  | { success: false; error: string }
+> {
   const openAIClient = createOpenAIClient({
     app: "lesson-assistant",
     chatMeta: {
       chatId,
       userId,
     },
   });
-  const responseFormat = zodTextFormat(
-    agent.schema,
-    `${agent.name}_response_schema`,
-  );
+  const schema = z.object({
+    output: z.union([
+      z.object({ success: z.literal(true), value: agent.schemaForLLM }),
+      z.object({ success: z.literal(false), error: z.string() }),
+    ]),
+  });
+  const responseFormat = zodTextFormat(schema, `${agent.name}_response_schema`);
 
-  const result = await openAIClient.responses.parse({
-    instructions: agent.prompt,
-    input: `### Examples from similar lessons
+  const instructions = agent.prompt;
+  const input = `### Examples from similar lessons
 
-${ragData.map((lesson) => `#### ${targetKey} ${lesson.title}\n\n${agent.extractRagData(lesson)}`).join("\n\n")}
+${ragData.map((lesson, i) => `#### Example ${i + 1}: ${targetKey} from lesson '${lesson.title}'\n\n${agent.extractRagData(lesson)}`).join("\n\n")}
  
     ### Document
 ${JSON.stringify(document, null, 2)}
 
 ### Additional Instructions
-${additionalInstructions}`,
+${additionalInstructions}`;
+
+  log.info(`Agent: ${agent.name}`);
+  log.info("Instructions:", instructions);
+  log.info("Input:", input);
+
+  const result = await openAIClient.responses.parse({
+    instructions,
+    input,
     stream: false,
     model: "gpt-4.1-2025-04-14",
     text: {
       format: responseFormat,
     },
   });
 
-  const parsedResult = result.output_parsed;
+  if (
+    result.output_parsed?.output?.success &&
+    "value" in result.output_parsed.output
+  ) {
+    const strictParseResult = agent.schemaStrict.safeParse(
+      result.output_parsed.output.value,
+    );
+    if (!strictParseResult.success) {
+      /**
+       * Retry logic in the case that the initial agent response is not valid.
+       * The reason we need this is that structured output does not support
+       * all the schema validation that we need.
+       */
+      const inputWithRetryInstructions =
+        input +
+        `\n\n### Note\nThis is a retry because the previous response was not valid. Your previous response output was: ${JSON.stringify(
+          result.output_parsed.output.value,
+        )}, however the strict schema is: ${JSON.stringify(
+          agent.schemaStrict,
+        )}; the error was ${JSON.stringify(strictParseResult.error.errors)} If the user's request is at odds with the schema, please respond with a message explaining the issue. Another agent will handle messaging to the user.`;
```

Comment:
```suggestion
        )}; the error was ${JSON.stringify(strictParseResult.error.errors)} If the user's request is at odds with the schema, please respond with an internal message explaining the issue. Another agent will handle messaging to the user.`;
```

Could this help clarify the message it's making?